src/main/java/app/linguistai/bmvp/service/stats/UserLoggedDateService.java

```diff
@@ -0,0 +1,75 @@
+package app.linguistai.bmvp.service.stats;
+
+import app.linguistai.bmvp.exception.NotFoundException;
+import app.linguistai.bmvp.model.User;
+import app.linguistai.bmvp.model.stats.UserLoggedDate;
+import app.linguistai.bmvp.repository.IAccountRepository;
+import app.linguistai.bmvp.repository.stats.IUserLoggedDateRepository;
+import app.linguistai.bmvp.response.stats.RUserLoggedDate;
+import app.linguistai.bmvp.utils.DateUtils;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.sql.Date;
+import java.time.LocalDate;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor
+@Service
+public class UserLoggedDateService {
+
+	private final IUserLoggedDateRepository userLoggedDateRepository;
+	private final IAccountRepository accountRepository;
+
+	public RUserLoggedDate getLoggedDates(String email, String sort, Integer numDays) {
+		// Validate numDays
+		if (numDays != null && numDays < 0) {
+			throw new IllegalArgumentException("Number of days must be a non-negative integer");
+		}
+
+		// Calculate startDate if numDays is not null
+		Date startDate = null;
+		if (numDays != null) {
```

Comment:
at this point we know `numDays` is not null (it is checked in line `30`), you can remove this check for more readability and better performance

Reply:
this would also make `startDate = ...` instead of `startDate = null`

Reply:
Can you check again? Because numDays **can** be null, it is an optional parameter of the related endpoint. Line 30 checks the case where numDays is not null and is negative, which is not allowed.

Reply:
@can-aslan

Reply:
Yes, you are correct. Maybe this could have been done differently to reduce the if checks but this implementation is good too in my opinion. I'm resolving this conversation, thank you for pointing it out 😄